### ISSUE TITLE
Demo June 2017

### ISSUE THREAD

**Issue by jeppe-style** [benchflow/benchflow#403]
Generic demo presentation idea:

- Open the `benchflow-test.yaml`, and the `docker-compose.yaml` when necessary, to describe the scenario and show how this is defined in the DSL;
- Submit the test using the APIs;
- Use the APIs to monitor the status of the experiment;
- Show artefact on Minio (generated test/experiment bundles, collected data, etc...) when relevant [you are going to get the question about the availability of a link in the API response]; 

Needs to be solved before next demo:
- [x] DSL: updates and missing parts (priorities) https://github.com/benchflow/benchflow/issues/396
- [x] DSL: enable `enumerations` use in Java https://github.com/benchflow/benchflow/issues/447 (fixed in PR https://github.com/benchflow/benchflow/pull/463)
- [x] DSL: step function https://github.com/benchflow/benchflow/issues/397 (fixed in PR https://github.com/benchflow/benchflow/pull/465)
- [x] DSL + TM: support for load test https://github.com/benchflow/benchflow/issues/402 (fixed in PR https://github.com/benchflow/benchflow/pull/468)
- [x] TM: update REST API response objects with links https://github.com/benchflow/benchflow/issues/445 (fixed in PR https://github.com/benchflow/benchflow/pull/469)
- [x] EM: update REST API response objects with links https://github.com/benchflow/benchflow/issues/446 (fixed in PR https://github.com/benchflow/benchflow/pull/470)
- [x] EM: adjust how results are retrieved from Faban https://github.com/benchflow/benchflow/issues/440 (fixed in PR https://github.com/benchflow/benchflow/pull/472)
- [x] FC + EM: illegal run status (unknown) https://github.com/benchflow/benchflow/issues/409 and PR https://github.com/benchflow/benchflow/issues/442 (fixed in PR https://github.com/benchflow/benchflow/pull/473)
- [x] TM: test termination criteria (max time) https://github.com/benchflow/benchflow/issues/337 (fixed in PR https://github.com/benchflow/benchflow/pull/512)
- [x] DSL + TM: enable storage of exploration space in DB https://github.com/benchflow/benchflow/issues/391 (fixed for DSL in PR https://github.com/benchflow/benchflow/pull/390) (fixed for TM in PR https://github.com/benchflow/benchflow/pull/476)
- [x] TM: update REST API for exploration point https://github.com/benchflow/benchflow/issues/418 (depends on #391) (fixed in PR https://github.com/benchflow/benchflow/pull/478)
- [x] EM: add REST API to abort running experiment (wait for trial to finish for now) (fixed in PR https://github.com/benchflow/benchflow/pull/479)

**General Issues:**
- [x] TM: return exploration space as part of status (issue https://github.com/benchflow/benchflow/issues/510) (3h) (fixed in PR https://github.com/benchflow/benchflow/pull/519)
- [x] TM: test status doesn't serialize max running time properly (issue https://github.com/benchflow/benchflow/issues/502) (1h) (fixed in PR https://github.com/benchflow/benchflow/pull/520)
- [x] EM: experiment status - collect the faban related info into one object and store it for each execution (issue https://github.com/benchflow/benchflow/issues/503) (1h)
- [x] TM: consider the user point of view in responses, e.g., also work if software client (links can be used directly, no need to parse information, needs more specification) (1h)
- [x] TM: throws exception when receiving experiment state if test is TERMINATED (issue https://github.com/benchflow/benchflow/issues/501) (0.5 h) (fixed in PR https://github.com/benchflow/benchflow/pull/521)
- [x] EM: If there is an error in the start task with Driver-Maker or Faban-Manager interactions then the experiment should go directly to the error state (issue https://github.com/benchflow/benchflow/issues/504) (2h)
- [x] TM + EM: retry if there are errors with minio connection (3 times)  (resilience) (issue https://github.com/benchflow/benchflow/issues/505, and related to https://github.com/benchflow/benchflow/issues/300) (1h) (fixed in PR https://github.com/benchflow/benchflow/pull/530 and PR https://github.com/benchflow/benchflow/pull/531

**Use Case 5 Issues:**
- [x] DSL: assert that mixes are correctly converted to previous DSL version (specifically `%`) (issue https://github.com/benchflow/benchflow/issues/508) (0.5 h)

**Use Case 7 Issues:**
- [x] TM: sometimes the (in use case 7) not all experiments are generated (issue https://github.com/benchflow/benchflow/issues/506) (3h)

**Use Case 9 Issues:**
- [x] TM: add new state `TERMINATING` where we wait when a test is about to be terminated but we still need to get an update from the EM about the currently running experiment (issue https://github.com/benchflow/benchflow/issues/499) (2h)
- [x] EM: task scheduler is locked on running experiment since we have polling => cannot abort experiment since lock is held by another thread (issue https://github.com/benchflow/benchflow/issues/500) (3h)

**DSL Improvements:**
- [x] DSL: add initial semantic check (issues https://github.com/benchflow/benchflow/issues/400 and https://github.com/benchflow/benchflow/issues/48) (1h) (done in PR https://github.com/benchflow/benchflow/pull/551)
- [ ] DSL: remove the constraints on the range order and complete step function implementation (see https://github.com/benchflow/benchflow/issues/397 and https://github.com/benchflow/benchflow/issues/397#issuecomment-314408192) (2h)

**API Additions**
- [x] TM: abort API that terminates the test calling terminate (see  thesis) (fixed in PR https://github.com/benchflow/benchflow/pull/554)
- [x] TM: list all tests of a user (see thesis) (fixed in PR https://github.com/benchflow/benchflow/pull/556)

**Other Issues:**
- [x] NEEDED FOR FAST CHECK THE POINTs ABOVE: End-to-end Tests https://github.com/benchflow/benchflow/issues/498
- [x] ALL: Everything in `Needs to be solved before next demo` that has still open issues
- [ ] ALL: Everything in #467 section `Jesper (@jeppe-style)`

**Boundary First Exploration**
- [ ] DSL + TM: add stability boundary exploration (issue https://github.com/benchflow/benchflow/issues/507, and related to #287 and the fact we need to store the decisions behind selecting specific points) (3h)
- [ ] DSL + TM: add boundary first exploration (issue https://github.com/benchflow/benchflow/issues/509 in progress...) (5h)

Provide Demo of the following (all WfMS):
**Basic Goal Use Cases**
- [x] **01** - _single test (load) success_ (real execution, after test supporting it)
  - requires: Adjust Goal Types and Update Life Cycle Handling https://github.com/benchflow/benchflow/issues/402
  - motivation: show that single experiments are supported (basic goal)
  - presentation: show successful execution using `GET /status`

**Goal Use Cases**
- [x] **02** -  _successful exhaustive exploration single variable - number of users_ (real execution, after test supporting it)
  - settings: 
    - one-at-a-time selection strategy
    - exploration space number of users
  - motivation: show that exhaustive exploration goal with workload/experiment generation is 
supported
  - presentation: show successful execution using `GET /status`
- [x] **03** - _successful exhaustive exploration single variable - memory_ (real execution, after test supporting it)
  - settings: 
    - one-at-a-time selection strategy
    - exploration space memory
  - motivation: show that exhaustive exploration goal with experiment and deployment descriptor generation is supported
   - presentation: show successful execution using `GET /status`, show `docker-compose` changed on minio
   - **note**:  bug in driver-maker
- [x] **04** - _successful exhaustive exploration with random selection - number of users_ (real execution, after test supporting it)
  - settings: 
    - random breakdown selection strategy
    - exploration space number of users (different to ensure not stored knowledge)
  - motivation: show that exhaustive exploration goal with random selection strategy is supported and 
give same result as one-at-a-time but different order (requirement for prediction)
  - presentation: show successful execution using `GET /status` and that order is random
- [x] **05** - _exhaustive exploration with failures - number of users_  (real execution, after test supporting it)
  - settings:
    - one-at-a-time selection strategy
    - exploration space number of users (50% known to overload system)
  - motivation: show that BenchFlow can handle SUTs that fail from client point of view, important for boundary exploration
  - VF: maybe you can also have a spike load function to stress the system and introduce failures
  - presentation: show successful execution using `GET /status` with failed experiments
- [x] **06** - _exhaustive exploration with failures - number of users_ (experiment-manager in test-mode)
  - settings:
    - one-at-a-time selection strategy
    - exploration space number of users
    - experiment-manager in test mode with scenario to fail every second experiment 
  - motivation: show that the experiment-manager works in test-mode and that it is useful for checking functionality of the system
  - presentation: show successful execution using `GET /status` and that every second experiment has failed
- [x] **07** - _2D exploration: number of users and DB initial memory allocation pool_ (real execution, after test supporting it) 
  - settings:
    - one-at-a-time selection strategy
    - exploration space: 
      - number of users 
      - DB `JAVA_OPTS` variable with values `["-Xms128m", "-Xms512m", "-Xms1024m"]`
  - motivation: show that multiple dimension exploration works (it also takes long time so we need prediction)
  - presentation: show successful execution using `GET /status`, show that `experiment definition`s are different, and that `docker-compose` files are different (per experiment)
- [x] **09** - exhaustive exploration (success) with **max time for test**(real execution): 
  - settings:
    - one-at-a-time selection strategy
    - exploration space: number of users (10 experiments)
    - max time 2 min
  - requires https://github.com/benchflow/benchflow/issues/337
  - motivation: show that max execution time is supported and that test has state `terminated: partially complete` and last experiment `terminated: aborted`
- [x] **10** - stepwise increasing user workload (different step functions, linear, exponential): 
    - settings:
        - one-at-a-time selection strategy
        - exploration space: number of users => range: [50,175], step: "+50" => 50, 100, 150, 175
    - requires https://github.com/benchflow/benchflow/issues/397
    - motivation: show that range can be specified in multiple ways (+, * ^)
    - presentation: show successful execution using `GET /status` and the exploration space
- [ ] **08** - exhaustive exploration (success) with **reuse of stored knowledge** (real execution): 
  - settings:
    - one-at-a-time selection strategy
    - exploration space: number of users (10 experiments)
    - ensure some (not all) experiments have been executed
  - requires: https://github.com/benchflow/benchflow/issues/381
  - motivation: show that reusing stored knowledge improves overall execution time
  - **note**: not implemented

**Performance Engineering Activities**
see https://github.com/benchflow/benchflow/issues/403#issuecomment-308070469

**Comment by VincenzoFerme** [comment]
@jeppe-style use headings and section to describe the scenario you plan to showcase more in depth and what you plan to explain out of the thesis work in that scenario. Then also describe which ones are supposed to be real executions and which are not, and which are the cases you cover in tests, that you can also showcase.

**Comment by VincenzoFerme** [comment]
Try to integrate also the other strategies we discussed to implement (e.g., boundary first) in the scenarios and a couple on other use cases using more configuration variables, in different contexts and with different data types (e.g., numbers, strings etc etc...)

Moreover, we need also a couple of scenario where you actually show the importance of the automation for specific tasks you want to perform out of the obtained data (but that are probably not yet integrated as a goal, so they can be future work integrations). For example:

- Resource sizing: perform experiments so that we can reproduce (with automation) something similar to what we did in the following paper, even though it is not automated: https://link.springer.com/chapter/10.1007/978-3-319-45468-9_5
- Before and After: Measure interval overlap with increasing user workload (different step functions, linear, exponential or adaptive to get good precision) - see the work done in: https://benchflow.slack.com/files/fermevincenzo/F5S8F0W2V/2017_benchflow_bpm_execution_unaware_processes_full__then_journal_.pdf

**Comment by VincenzoFerme** [comment]
@jeppe-style Some notes (that have to be converted to dedicated issues when ready to be solved) out of the execution of most of the scenario above:

1. We need to order the generation of the exploration space, so that it is deterministic the mapping between the order of the specification in the YAML to the order of the points in the `one-at-a-time` strategy
2. ~~Example 1: Validate termination criteria (as running state) -> lastRunningState (now: currentRunningState)~~
3. ~~DSL termination criteria, experiment, type: ==> are they clear enough?~~
4. Extend to generate the other conditions as well in the test [also the ones that are not supported for execution, so that is generic]
5. Example 6: make it clear that there is a severe failure in Faban (maybe with Scenario name or something else) + update the name so that is clear what it does
6. Check why we get the same data back for tests with same names + why it is only one...
7. Check why `JAVA_OPTS=` was not added to the `db` service in the generated `docker-compose` file
8. Have more realistic examples, and apply those examples also to the defined tests

**Comment by jeppe-style** [comment]
2\. _Example 1: Validate termination criteria (as running state) -> lastRunningState (now: currentRunningState)_ => it is currently `runningState`, and we agreed that this was fine because it is not the lastRunningState
3\. _DSL termination criteria, experiment, type: ==> are they clear enough?_ => I thought that was ok once explained
4\. _Extend to generate the other conditions as well in the test [also the ones that are not supported for execution, so that is generic]_ => I don't know what this means
6\. _Check why we get the same data back for tests with same names + why it is only one..._ => I don't know what this means
8\. _Have more realistic examples, and apply those examples also to the defined tests_ => in which way are the current examples not realistic?

**Comment by VincenzoFerme** [comment]
4\. So, I think this apply to the Termination Criteria, that how is somehow hardcoded.
6\. When you executed different tests having the same name, you where getting wrong data as result in the REST API response
8\. We discussed it in: https://benchflow.slack.com/archives/D5DTWPVML/p1497270804890918, https://benchflow.slack.com/archives/G5GCXP63D/p1497426264804953. It is mainly about having realistic values as settings for the exploration space.

**Comment by VincenzoFerme** [comment]
Some behaviours to be fixed experienced during the last demo (please open fine grain issues for each of them before working on a fix):

1. Mapper for maxRunningTime
2. Empty trial, need to send info back to the test manager, because now it is only in the experiment manager
3. Make the experiment manager not to stop in a waiting state outside of the queue
4. If error in start then it should go in the Error state (EM)
5. After a fast kill the status was empty for the experiment


- Scenario 7

=> Have resilient connections and acks in the interaction with other services, otherwise it is really difficult to understand the causes of the issues (see: https://github.com/Financial-Times/resilient-jersey-wrapper and https://github.com/jhalterman/failsafe + (theory) http://www.javaworld.com/article/2824163/application-performance/stability-patterns-applied-in-a-restful-architecture.html)
=> We had issues that we had missing experiments/trial for one exploration point, understand what is going on

- Scenario 9

=> Update with Partial complete state

**Comment by jeppe-style** [comment]
@VincenzoFerme I have updated the first comment in this issue with `issues found` there I list the items I plan to work on to get all working.
1. it is in the list on top
2. I don't think this information is necessary at the moment. I suggest to remove this from the status until there is an actual need.
3. I'm not sure what this is about
4. moved to the top
5. Will probably be fixed with https://github.com/benchflow/benchflow/issues/499

added issues for resilience connection with Minio on top.
